docs/proposals/devfile/outer-loop-build-and-deploy.md

```diff
@@ -0,0 +1,310 @@
+          component: mydockerfileimage
+
+__Note:__
+1. The `apply` command with the `image` component is optional. If the `apply` command that references an image component does not exist, then the image build will be done at the startup automatically (similar to the behaviour of the existing `kubernetes` components).  If there is an `apply` command that references the `image` component, then the user will need to create a composite command to chain the image build apply command with the deploy command (see the deployment command section) to complete the build and deploy.
+ 
+### Example using a Dockerfile stored in a git repo with push registry without apply command:
+    variables:
+      myimage: myimagename
+    components:
+      - name: mydockerfileimage
+        image:
+          imageName: {{myimage}}
+          dockerfile:
+            git:
+              remotes:
+                origin: "https://github.com/odo-devfiles/nodejs-ex.git"
+            location: Dockerfile
+            args: [ ]
+
+The git definition will be the same as the one in `starterProjects` definition that supports `checkoutFrom`
+
+`location`: the location of the dockerfile within the repo. If `checkoutFrom` is being used, the location will be relative to the root of the resources after cloning the resources using the `checkoutFrom` setting.
+
+Notes:
+1. The build tool/mechanism, e.g. buildah/kaniko, used for building the dockerfile is up to the tools so it is not part of the spec
+1. If endpoints definition is needed, it will be defined via the corresponding deployment manifest in the deployment step, e.g. inside the kubernetes deployment manifest.
+1. Do we need image push registry info to specify where to push the image to? If needed, then the secret for the push can be specified as `secretKeyRef` in https://github.com/devfile/api/issues/299
+
+### Example using SourceToImage (S2I):
+    variables:
+      myimage: myimagename
+    components:
+      - name: mys2iimage
+        image:
+          imageName: {{myimage}}
+          s2i:
+            builderImageNamespace: mynamespace
+            builderImageStreamTag: mytag
+            scriptLocation:
+              remotes:
+                origin: "https://github.com/odo-devfiles/nodejs-ex.git"
+    commands:
+      - id: deploybuild
+      apply:
+        component: mys2iimage
+
+`builderImageNamespace`: Namespace where builder image is present
+
+`builderImageStreamTag`: Builder image name with tag
+
+`scriptionLocation`: Script URL to override default scripts provided by builder image
+
+`incrementalBuild`: Flag that indicates whether to perform incremental builds. Default is true (optional)
+
+## Deployment of the image
+Deploy the image to a cluster. Different technologies can be used for deploying the image, for example:
+1. Kubernetes deployment manifest
+1. Operators
+1. Helm
+
+The initial design will focus on the first two items for discussion. Other ways of building the image will be covered later.
```

Comment:
These items may need to be reordered -- it looks like the proposal below covers deployment manifest and helm.

Reply:
I'll reorder the example to match.